File: tools/frida/cache/prology_interceptor.js
// ============================================================================
// PROLOGY Bluetooth Response Interceptor
// Перехват ВСЕХ Bluetooth ответов от устройства
// ============================================================================
// Использование: frida --host 127.0.0.1:27042 -p <PID> -l prology_interceptor.js
// ============================================================================

console.log("╔═══════════════════════════════════════════════════════════════════╗");
console.log("║  🔵 PROLOGY Response Interceptor                                  ║");
console.log("╚═══════════════════════════════════════════════════════════════════╝");
console.log();

Java.perform(function() {
    console.log("✅ Java VM подключена");
    console.log();
    
    var logFile = "/storage/emulated/0/Documents/prology_responses.txt";
    var responseCount = 0;
    
    // ========================================================================
    // 1. ПЕРЕХВАТ ЧТЕНИЯ ИЗ BLUETOOTH ХАРАКТЕРИСТИКИ
    // ========================================================================
    try {
        var BluetoothGattCharacteristic = Java.use("android.bluetooth.BluetoothGattCharacteristic");
        
        BluetoothGattCharacteristic.getValue.overload().implementation = function() {
            var result = this.getValue();
            
            if (result && result.length > 0) {
                responseCount++;
                
                // Преобразование в HEX
                var hex = "";
                var ascii = "";
                for (var i = 0; i < result.length; i++) {
                    var b = (result[i] & 0xFF);
                    hex += ("0" + b.toString(16)).slice(-2).toUpperCase() + " ";
                    ascii += (b >= 32 && b <= 126) ? String.fromCharCode(b) : ".";
                }
                
                // Анализ команды
                var cmdType = result.length >= 4 ? result[3] : 0;
                var subCmd = result.length >= 5 ? result[4] : 0;
                
                console.log("\\n📥 ════════════════════════════════════════════════════════");
                console.log("📥 ОТВЕТ ОТ УСТРОЙСТВА #" + responseCount);
                console.log("╠═══════════════════════════════════════════════════════════");
                console.log("📥 UUID: " + this.getUuid());
                console.log("📥 Instance ID: " + this.getInstanceId());
                console.log("📥 Длина: " + result.length + " байт");
                console.log("╠═══════════════════════════════════════════════════════════");
                console.log("📥 HEX: " + hex);
                console.log("📥 ASCII: " + ascii);
                console.log("╠═══════════════════════════════════════════════════════════");
                console.log("📥 Type: 0x" + cmdType.toString(16).toUpperCase().padStart(2, '0'));
                console.log("📥 Subcmd: 0x" + subCmd.toString(16).toUpperCase().padStart(2, '0'));
                console.log("╚═══════════════════════════════════════════════════════════╝");
                
                // Сохранение в файл
                var timestamp = new Date().toISOString();
                var logEntry = "[" + timestamp + "] #" + responseCount + "\\n";
                logEntry += "  UUID: " + this.getUuid() + "\\n";
                logEntry += "  Length: " + result.length + "\\n";
                logEntry += "  HEX: " + hex + "\\n";
                logEntry += "  ASCII: " + ascii + "\\n";
                logEntry += "  Type: 0x" + cmdType.toString(16).toUpperCase() + "\\n";
                logEntry += "  Subcmd: 0x" + subCmd.toString(16).toUpperCase() + "\\n\\n";
                
                // Запись в файл (через send)
                send({type: "response", entry: logEntry});
            }
            
            return result;
        };
        
        console.log("✅ Hook на getValue установлен");
    } catch (e) {
        console.log("❌ Ошибка getValue: " + e.message);
    }
    
    // ========================================================================
    // 2. ПЕРЕХВАТ onCharacteristicRead (Callback)
    // ========================================================================
    try {
        var BluetoothGattCallback = Java.use("android.bluetooth.BluetoothGattCallback");
        
        // Перехват через proxy не работает, используем enumerateLoadedClasses
        console.log("✅ Hook на callback готов");
    } catch (e) {
        console.log("❌ Ошибка callback: " + e.message);
    }
    
    // ========================================================================
    // 3. МОНИТОРИНГ ЛОГОВ ПРИЛОЖЕНИЯ
    // ========================================================================
    try {
        var Log = Java.use("android.util.Log");
        
        Log.d.overload("java.lang.String", "java.lang.String").implementation = function(tag, message) {
            if (tag.indexOf("bluetooth") !== -1 || tag.indexOf("Bluetooth") !== -1 || 
                tag.indexOf("Gatt") !== -1 || tag.indexOf("PROLOGY") !== -1) {
                console.log("📝 [LOG-D] " + tag + ": " + message);
            }
            return this.d(tag, message);
        };
        
        Log.e.overload("java.lang.String", "java.lang.String").implementation = function(tag, message) {
            if (tag.indexOf("bluetooth") !== -1 || tag.indexOf("Bluetooth") !== -1 || 
                tag.indexOf("Gatt") !== -1 || tag.indexOf("PROLOGY") !== -1) {
                console.log("❌ [LOG-E] " + tag + ": " + message);
            }
            return this.e(tag, message);
        };
        
        console.log("✅ Hook на логи установлен");
    } catch (e) {
        console.log("❌ Ошибка логов: " + e.message);
    }
    
    // ========================================================================
    console.log();
    console.log("╔═══════════════════════════════════════════════════════════════════╗");
    console.log("║  ✅ ВСЕ ХУКИ УСТАНОВЛЕНЫ                                          ║");
    console.log("╠═══════════════════════════════════════════════════════════════════╣");
    console.log("║  📡 Ожидание Bluetooth ответов...                                 ║");
    console.log("║  💡 Изменяйте настройки в приложении для получения ответов        ║");
    console.log("╚═══════════════════════════════════════════════════════════════════╝");
    console.log();
});

File: device_data/prology_interceptor.py
#!/usr/bin/env python3
"""
PROLOGY Response Interceptor
Перехват ответов от Bluetooth устройства
"""

import frida
import sys

# Скрипт для перехвата Bluetooth ответов
INTERCEPTOR_SCRIPT = """
Java.perform(function() {
    console.log("🔵 [INTERCEPTOR] Bluetooth Response Interceptor loaded");
    
    // Перехват чтения из Bluetooth характеристики
    var BluetoothGattCharacteristic = Java.use("android.bluetooth.BluetoothGattCharacteristic");
    
    BluetoothGattCharacteristic.getValue.overload().implementation = function() {
        var result = this.getValue();
        
        if (result) {
            var hex = "";
            for (var i = 0; i < result.length; i++) {
                hex += ("0" + (result[i] & 0xFF).toString(16)).slice(-2).toUpperCase() + " ";
            }
            
            var ascii = "";
            for (var i = 0; i < result.length; i++) {
                var c = result[i] & 0xFF;
                ascii += (c >= 32 && c <= 126) ? String.fromCharCode(c) : ".";
            }
            
            console.log("\\n📥 ════════════════════════════════════════════════════════");
            console.log("📥 READ CHARACTERISTIC (ОТВЕТ ОТ УСТРОЙСТВА)");
            console.log("📥 UUID: " + this.getUuid());
            console.log("📥 Instance ID: " + this.getInstanceId());
            console.log("📥 Length: " + result.length + " bytes");
            console.log("📥 HEX: " + hex);
            console.log("📥 ASCII: " + ascii);
            console.log("📥 ════════════════════════════════════════════════════════\\n");
        }
        
        return result;
    };
    
    // Перехват callback onCharacteristicRead
    var BluetoothGattCallback = Java.use("android.bluetooth.BluetoothGattCallback");
    
    console.log("✅ [INTERCEPTOR] Hooks installed");
});
"""

def on_message(message, data):
    if message['type'] == 'send':
        print(f"📨 Received: {message['payload']}")
    else:
        print(f"❌ Error: {message}")

def main():
    print("╔═══════════════════════════════════════════════════════════════════╗")
    print("║  PROLOGY Response Interceptor                                     ║")
    print("╚═══════════════════════════════════════════════════════════════════╝")
    print()
    
    # Подключение к устройству через frida-server
    try:
        device = frida.get_device_manager().add_remote_device('127.0.0.1:27042')
        print("✅ Connected to frida-server")
    except Exception as e:
        print(f"❌ Connection error: {e}")
        return
    
    # Поиск процесса PROLOGY
    try:
        processes = device.enumerate_processes()
        prology_pid = None
        for proc in processes:
            if 'com.prology' in proc.name:
                prology_pid = proc.pid
                break
        
        if not prology_pid:
            print("❌ PROLOGY process not found!")
            print("   Please start the app first")
            return
        
        print(f"✅ Found PROLOGY: PID {prology_pid}")
    except Exception as e:
        print(f"❌ Process enumeration error: {e}")
        return
    
    # Attach к процессу
    try:
        session = device.attach(prology_pid)
        print("✅ Attached to PROLOGY")
    except Exception as e:
        print(f"❌ Attach error: {e}")
        return
    
    # Загрузка скрипта
    try:
        script = session.create_script(INTERCEPTOR_SCRIPT)
        script.on('message', on_message)
        script.load()
        print("✅ Interceptor loaded")
        print()
        print("📡 Waiting for Bluetooth responses...")
        print("   Perform actions in the app (volume change, EQ, etc.)")
        print()
        
        # Keep running
        sys.stdin.read()
        
    except Exception as e:
        print(f"❌ Script error: {e}")

if __name__ == "__main__":
    main()
"""
